### PR TITLE
Add explicit dependency to validatePom and generatePom tasks.

### PR DESCRIPTION
**Pull request by zelinh**
### Description
This PR is to add explicit task dependency on PomGenerate and PomValidate tasks for different publication types.

 For plugin with different publication, all `generatePom` tasks such as `generatePomFileForNebulaPublication` or `generatePomFileForPluginZipPublication` are creating the Pom file at the same cache location: build/distributions/*******.pom; and Pom validation tasks are reading the output of this Pom. 

We will get error like this in `job-scheduler` plugin:
```
> Task :generatePomFileForPluginZipPublication FAILED

FAILURE: Build failed with an exception.

* What went wrong:
A problem was found with the configuration of task ':generatePomFileForPluginZipPublication' (type 'GenerateMavenPom').
  - Gradle detected a problem with the following location: '/Users/zelinhao/workplace/job-scheduler/build/distributions/opensearch-job-scheduler-2.12.0.0-SNAPSHOT.pom'.
    
    Reason: Task ':validateNebulaPom' uses this output of task ':generatePomFileForPluginZipPublication' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
    
    Possible solutions:
      1. Declare task ':generatePomFileForPluginZipPublication' as an input of ':validateNebulaPom'.
      2. Declare an explicit dependency on ':generatePomFileForPluginZipPublication' from ':validateNebulaPom' using Task#dependsOn.
      3. Declare an explicit dependency on ':generatePomFileForPluginZipPublication' from ':validateNebulaPom' using Task#mustRunAfter.
    
    For more information, please refer to https://docs.gradle.org/8.5/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.
```
**Reason**
Gradle treats any task that reads a file as implicitly dependent on each task that writes to that file.
In our case, Gradle is assuming each validation task is depending on all the Pom generation tasks as implicit dependency. 

In this PR, we are adding these implicit dependencies for `generatePomFile` and `validatePom` into its implementation. 

We can see same logic is added in few other plugin repos:
In [geospatial](https://github.com/opensearch-project/geospatial/blob/430077dc11ec2d77f9ba2c0f4e4f60fae8a3def3/libs/h3/build.gradle#L77): `validateNebulaPom.dependsOn(generatePomFileForPluginZipPublication)`
In [skills](https://github.com/opensearch-project/skills/blob/main/build.gradle#L185): `tasks.validateNebulaPom.dependsOn ':generatePomFileForPluginZipPublication'`

### Check List
~~- [x] New functionality includes testing.~~
~~  - [x] All tests pass~~
~~- [x] New functionality has been documented.~~
~~  - [x] New functionality has javadoc added~~
~~- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))~~

- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — APPROVED** [review]

**Comment by reta** [comment]
Thanks @zelinh , I think it looks good!

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@zelinh could you please backport to `2.x` manually? thank you

**Comment by cwperks** [comment]
@reta @zelinh I'm not sure yet, but I think this PR is contributing to failures being seen in security plugin's CI: https://github.com/opensearch-project/security/pull/4156

```
* Where:
Build file '/home/runner/work/security/security/build.gradle' line: 417

* What went wrong:
A problem occurred evaluating root project 'opensearch-security'.
> NamedDomainObjectContainer#configureEach(Action) on Publication container cannot be executed in the current context.
```

**Comment by reta** [comment]
Ah, that is unexpected, thanks a lot for @cwperks !

File: buildSrc/src/main/java/org/opensearch/gradle/precommit/PomValidationPrecommitPlugin.java [hunk 1]
```diff
@@ -48,17 +48,24 @@ public class PomValidationPrecommitPlugin extends PrecommitPlugin {
     public TaskProvider<? extends Task> createTask(Project project) {
         TaskProvider<Task> validatePom = project.getTasks().register("validatePom");
         PublishingExtension publishing = project.getExtensions().getByType(PublishingExtension.class);
-        publishing.getPublications().all(publication -> {
+        publishing.getPublications().configureEach(publication -> {
```
Comment (cwperks) [inline buildSrc/src/main/java/org/opensearch/gradle/precommit/PomValidationPrecommitPlugin.java:51]: This line seemed to cause a breaking change in the security plugin:

```
* Where:
Build file '/home/runner/work/security/security/build.gradle' line: 417

* What went wrong:
A problem occurred evaluating root project 'opensearch-security'.
> NamedDomainObjectContainer#configureEach(Action) on Publication container cannot be executed in the current context.
```

Is this change from `all` -> `configureEach` necessary?

**Pull request merged**